### PR TITLE
Take submount etag into account for folder etags

### PR DESCRIPTION
**Pull request by icewind1991**
This removes the need to do cross-storage etag propagation

### CODE REVIEW

**Comment by icewind1991** [comment]
@nickvergessen can you smash this?

**Comment by nickvergessen** [comment]
Smashed: :white_check_mark:

**Comment by icewind1991** [comment]
I think we can try to get this in as it is currently, there is still some optimizations we can do on top of this (remove the files_external propagation logic) but in current state it should improve the performance without regressions.

cc @PVince81

**Comment by PVince81** [comment]
@icewind1991 do you mind writing a more detailed explanation of the etag propagation change as discussed last week ?

I'd also like to know what @DeepDiver1975 thinks.

While I agree that this brings a perf improvement and simplifies the etag propagation code, I'm a bit worried what happens if one user has 100+ received shared folders (mount points) and for every PROPFIND the computed etag needs to compute all of them.

File: lib/private/files/cache/propagator.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OC\Files\Cache;
+
+class Propagator {
```
Comment (PVince81) [inline lib/private/files/cache/propagator.php]: PHPDoc ? What does this propagator propagate and how ?

File: lib/private/files/view.php [hunk 2]
```diff
@@ -687,14 +687,14 @@ public function rename($path1, $path2) {
 					} else {
 						$result = false;
 					}
-				// moving a file/folder within the same mount point
+					// moving a file/folder within the same mount point
```
Comment (PVince81) [inline lib/private/files/view.php]: oops?

File: lib/private/files/view.php [hunk 3]
```diff
@@ -1164,6 +1160,43 @@ public function hasUpdated($path, $time) {
 	}
 
 	/**
+	 * @param \OC\Files\Storage\Storage $storage
+	 * @param string $internalPath
+	 * @param string $relativePath
+	 * @return array|bool
+	 */
+	private function getCacheEntry($storage, $internalPath, $relativePath) {
```
Comment (PVince81) [inline lib/private/files/view.php]: What does this method do ? Looks more complicated than a simple "get"

**Comment by PVince81** [comment]
Raised https://github.com/owncloud/core/issues/20521 to provide Webdav-based etag propagation tests, as discussed last week.

**Comment by icewind1991** [comment]
> While I agree that this brings a perf improvement and simplifies the etag propagation code, I'm a bit worried what happens if one user has 100+ received shared folders (mount points) and for every PROPFIND the computed etag needs to compute all of them.

Modern hardware has no problem calculating tens of thousands md5 hashes a second for strings of a few hundred characters, any overhead from the hashing will be insignificant compared to the other overhead that comes with 100+ shares/mounts

**Comment by PVince81** [comment]
and if I understand well, the mounts are already resolved and the resolution is something we already did in the past, so there should be no performance regression there ?

**Comment by icewind1991** [comment]
> and if I understand well, the mounts are already resolved and the resolution is something we already did in the past, so there should be no performance regression there ?

Correct

**Comment by icewind1991** [comment]
> do you mind writing a more detailed explanation of the etag propagation change as discussed last week ?

Currently when propagating etags (and mtimes) we need to propagate all the way up to the root for *all* users who have access to the file which is complicated and expensive.

This changes it to only propagate to the root of the storage (for the owner of the file) making the etags storage in the database storage-specific etags.

When reading the etag for a folder in `getFileInfo`/`getDirectoryContent` it calculates the etag for the folder based on the "storage etag" for the folder and the etag of the root of all mounts in the folder, thus whenever a file in a mounted storage changes the etag get's propagated to the root of that storage which will change the calculated etag for the parent folders of those mounts.

**Comment by PVince81** [comment]
Smashbox tests all passed.

**Comment by PVince81** [comment]
I'm not sure whether we should merge it now or wait for @SergioBertolinSG's Webdav level etag propagation tests ? @DeepDiver1975 what do you think ?

On the other hand having it merged early gives an opportunity for more developer testing.

@icewind1991 should this be "To review" or is anything open ?

**Comment by icewind1991** [comment]
This is good to merge in my opinion

**Comment by PVince81** [comment]
I'm ok with this :+1:

**Comment by nickvergessen** [comment]
> Currently when propagating etags (and mtimes) we need to propagate all the way up to the root for all users who have access to the file which is complicated and expensive.

The only problem with this is, that you change it from "write" to "read"? Now we need to always resolve this when reading the etag, as opposed to only when we change something, which happens way less?

**Comment by icewind1991** [comment]
> The only problem with this is, that you change it from "write" to "read"?

Yes, but the logic we need to do on read is simple compared to the *very* complex logic we had on write and the cost is neglectable

**Comment by PVince81** [comment]
Would be good to have integration tests too https://github.com/owncloud/core/issues/20521

**Comment by icewind1991** [comment]
@nickvergessen @DeepDiver1975 can this be merged?

**Comment by PVince81** [comment]
Not sure, I think I'd prefer to have the tests done first: https://github.com/owncloud/core/issues/20521 ?

But if it's blocking other work then I'm fine with merging it, considering that the tests will arrive in 9.0 too.

**Comment by icewind1991** [comment]
It's blocking further work on the storage api

**Comment by PVince81** [comment]
@icewind1991 do we still need this etag propagator for files_external ? https://github.com/owncloud/core/blob/v8.2.0/apps/files_external/lib/etagpropagator.php

**Comment by icewind1991** [comment]
No, but I prefer to have this merged as is before going trough all other etag propagation paths

**Comment by PVince81** [comment]
- [ ] REQUIRES: smashbox tests https://github.com/owncloud/smashbox/pull/136

**Comment by PVince81** [comment]
@DeepDiver1975 can we merge this ? The smashbox tests are **almost** ready but depend on the sync client version fix for owncloudcmd.

**Comment by nickvergessen** [comment]
The smashbox test has been merged.

**Comment by icewind1991** [comment]
@DeepDiver1975 @nickvergessen can the be merged?

**Comment by nickvergessen** [comment]
I will smash a fresh master with this being merged in, so we can test it with all the latest magic,
but fine by me

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by nickvergessen** [comment]
Smashed again and passed: :white_check_mark: :+1:

**Pull request merged**